src/src/com/microsoft/aad/adal/AcquireTokenSilentHandler.java

```diff
@@ -0,0 +1,359 @@
+        }
+        
+        mContext = context;
+        mAuthRequest = authRequest;
+        mTokenCacheAccessor = tokenCacheAccessor;
+
+        mWebRequestHandler = new WebRequestHandler();
+    }
+    
+    /**
+     * Request for access token by looking up cache. . 
+     */
+    AuthenticationResult getAccessToken() throws AuthenticationException {
+        // If mTokenCacheAccessor is null, won't handle with token cache lookup. 
+        if (mTokenCacheAccessor == null) {
+            return null;
+        }
+        
+        TokenCacheItem tokenCacheItem = mTokenCacheAccessor.getRegularTokenCacheItemWithAT(mAuthRequest.getResource(), 
+                mAuthRequest.getClientId(), getUser());
+        
+        if (tokenCacheItem == null 
+                && lookUpCachedTokenWithNoUserForADFS()) {
+            // ADFS doesn't return idtoken back. If app is talking to ADFS server, and provide us the userId
+            // or displayableId, since we didn't get idtoken back, we don't store user as part of key in this
+            // case. Try again to find token with no user.
+            tokenCacheItem = mTokenCacheAccessor.getRegularTokenCacheItemWithAT(mAuthRequest.getResource(), 
+                    mAuthRequest.getClientId(), null);
+            // If we still don't have anything to use then we should try to find if we have a MRRT that matches
+            if (tokenCacheItem == null) {
+                return tryMRRT();
+            }
+        }
+        
+        if (isValidCache(tokenCacheItem)) {
+            if (isUserMisMatch(tokenCacheItem)) {
+                Logger.v(TAG, "User in retrieved token cache item does not match the user provided in the request.");
+                throw new AuthenticationException(ADALError.AUTH_FAILED_USER_MISMATCH);
+            }
+            
+            Logger.v(TAG, "Valid token cache item, return AT back.");
+            return AuthenticationResult.createResult(tokenCacheItem);
+        }
+        
+        return tryRT();
+    }
+    
+    /**
+     * Send token request with grant_type as refresh_token to token endpoint for getting new access token. 
+     */
+    AuthenticationResult acquireTokenWithRefreshToken(final String refreshToken) 
+            throws AuthenticationException {
+        Logger.v(TAG, "Try to get new access token with the found refresh token.", 
+                mAuthRequest.getLogInfo(), null);
+        
+        // @note: The original intention for checking network connection before sending RT request is to avoid RT removal
+        // from cache if user turned on airplane mode or similar case. However, the updated cache logic only remove RT when 
+        // receiving oauth2 error invalid_grant from server. 
+        // The only reason to keep it here is in case there is caller taking dependency on the below error code. 
+        final DefaultConnectionService connectionService = new DefaultConnectionService(mContext);
```

Comment:
pull out this code into a separate method throwIfConnectionNotAvaliable, put it into a central place and reuse.